add @appium/types package

boneskull opened the pull request:
This PR does two things:

1. Provides some type declarations which will be used across other packages and consumers in `@appium/types`
2. New TS configuration
    This changes the TS configuration to to a) emit declarations for supported packages, and b) build declarations incrementally by package.

    To begin, we are targeting `appium`, `@appium/base-driver` and `@appium/support` for declarations; these three are intended to be published with declarations generated via their JS.

    Both `@appium/support` and `appium` are fully typechecked (sans implicit `any` types), but `@appium/base-driver` is mostly not.  Regardless, declarations are generated for all three.

    Each package will have its own `tsconfig.json` which "inherits" from `config/tsconfig.base.json`.  Each will also need to declare relative paths to dependencies. e.g., the `appium` package must configure its `tsconfig.json` so that it depends on the declarations of `@appium/support` and also where to find `@appium/support`. Essentially, a small dependency tree gets built, and those at the "root" get their declarations built first; then the consumers use those declarations.

boneskull committed:
feat(types): add new @appium/types package

This changes the TS configuration to to a) emit declarations for supported packages, and b) build declarations incrementally by package.

To begin, we are targeting `appium`, `@appium/base-driver` and `@appium/support` for declarations; these three are intended to be published with declarations generated via their JS.

Both `@appium/support` and `appium` are fully typechecked (sans implicit `any` types), but `@appium/base-driver` is mostly not.  Regardless, declarations are generated for all three.

Each package will have its own `tsconfig.json` which "inherits" from `config/tsconfig.base.json`.  Each will also need to declare relative paths to dependencies. e.g., the `appium` package must configure its `tsconfig.json` so that it depends on the declarations of `@appium/support` and also where to find `@appium/support`. Essentially, a small dependency tree gets built, and those at the "root" get their declarations built first; then the consumers use those declarations.

Also reorganized some scripts and added `npm-run-all` to help and added a "typecheck" job to CI.

Since `npm install` will build, it's pretty inconvenient to have `npm` abort installation if the build fails.  In that case, if any of the build steps fail (see `build:loose`) the installation will continue.  Added a `prepublishOnly` which runs a build in "strict" mode where everything must pass.

_Note: there's still a strong coupling of "what `BaseDriver` provides" and "what an external driver *may* provide".  It's unclear to me if this is a problem, though there seem to be a handful of methods that external drivers *must* implement.  The easiest way to get around this would probably be to extract these required properties and methods into their own interface (duplication) if we cannot otherwise declare methods as "abstract".  As it stands, the `Driver` is basically just `BaseDriver`._

Also:

- Moved some configuration types out of `appium` and into here, since they are used both by `appium` and `base-driver` via `DriverOpts`
- Split `Driver` type into "the thing that BaseDriver is" and "all the other methods external drivers can implement"
- Added missing `proxyCommand` method
- Better generics for `LogType`
- Remove unused `Constructor` type
- Better types for `getSession`/`getSessions`
- Added util `Class` type
- Added many missing typings from DefinitelyTyped
- Added `ts-node` for `@wdio/types` (I need to send a PR to webdriver to eliminate this dependency)
- Type checks in their CI step

boneskull committed:
feat(support): generate declaration files

Most of these changes are the addition of types.  There are a few places in which the typechecker found weirdness, and I have addressed those.

- In `logging.js`, `parseInt()` was being called on props of `process.env`; this is not allowed, because `parseInt()` only accepts a string, and props may be `undefined`.
- In `log-internal.js`, I had to remove the `raiseError()` function because TypeScript is unable to understand that a function, when called, will always throw.  The fix was just to replace with `throw new Error...`
- Introduced a var in `mjpeg.js` to avoid repeated casting
- Added some type guards to `net.js`. as function parameters, union types kind of suck. This is definitely one place TS would have helped, since it allows function overloads.  Also dropped use of deprecated `url` module in lieu of W3C `URL`.
- Removed some old-node-version cruft from `util.js`.

Non-code changes:

- Needed to use an override for `type-fest`, as it's included as a transitive dependency, but otherwise is only present in the devdeps of the monorepo root `package.json`. This means, due to version conflicts, a `support/node_modules/type-fest` would be created with an old, incompatible version, which would be preferred over the monorepo one due to Node's resolution algorithm.  This needs npm v8.3+ to work properly, but is only a problem for contributors (not consumers).
- Pinned type of `opencvv-bindings`.
- No longer deleting `build` dir before transpilation since it may contain generated declarations!

boneskull committed:
feat(base-driver): generate declaration files

This was not as easy as the other two.  The main issue involves the `BaseDriver` class and is threefold:

1. `BaseDriver` and a valid external subclass thereof (a `Driver`) are not the same, because the former is essentially _abstract_. There is no notion of an abstract class in JS nor in TS for JS.
2. Splitting `BaseDriver` into several modules and "mixing in" some modules is just not well-understood by TS.
3. `AppiumDriver` inherits _most_ of `BaseDriver`, but overrides some functions with differing signatures. This is unsupported by TS.

To work around these problems, I had to:

1. Move the `Driver` interface into `@appium/types`, since other packages may depend on it.  We do not want circular dependencies, which is likely to occur otherwise.  Further, there is no way to tell TS for JS that a class _implements_ an interface or type.
2. Refactor the mixins (in `basedriver/commands/`) into "mixin classes", as understood by typescript. These are dynamic classes returned by functions which accept a class parameter, and _extend_ that class. The result is marginally better, but the implementation is ... well, see for yourself.  It is not possible to use declaration merging to just "trick" TS into thinking that `BaseDriver` has all of these extra methods; classes cannot be merged, and even if you could, you can't use declaration merging to "add" methods afaict.
3. `BaseDriver` is now an empty class which inherits from all mixins applied to `DriverCore`.  Due to syntactic restrictions, it must be a `class` statement in order to use the `@implements` tag, which causes TS to check that the result fully implements `Driver`.  `AppiumDriver` _also_ inherits from `DriverCore`, since it does not override `BaseDriver` "properly".  `BaseDriverCore` itself is pretty small, containing only those functions which `AppiumDriver` defines as differing.

Otherwise:

- `BaseDriver` does not inherit from `Protocol`, which is just an empty class. `Protocol` was removed.
- All of the error classes are exported from `protocol/errors`, because TS cannot see them otherwise.
- Exported missing `GET_STATUS_COMMAND` from `protocol/index`
- Re-exported `normalizeBasePath` from `server` instead of `protocol`, since that's where it's defined
- Removed `index.d.ts`
- Mixins use `@implements` as well. Note that `@implements` does not support `import('..')` syntax.

chore(base-driver): attempted to make types for mixins less awful

fixed a couple issues in the types in `driver.js`

core(base-driver): add capabilities types

- also move the bluebird config into the index
- fix references to `DriverOpts`
- add `assignServer()` to `BaseDriverCore`, used by `AppiumDriver`

chore(base-driver): fix more types

add missing reference to `@appium/types` in `tsconfig.json`

fix(base-driver): ensure getLogEvents is async

fix(base-driver): add some more type checks and fix logCustomEvent to be async

chore(base-driver): reformat

boneskull commented:
Current dependencies on/for this PR:
* 2.0
  * **PR #16594** <a href="https://app.graphite.dev/github/pr/appium/appium/16594" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  (this PR)
    * **PR #16595** <a href="https://app.graphite.dev/github/pr/appium/appium/16595" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
      * **PR #16596** <a href="https://app.graphite.dev/github/pr/appium/appium/16596" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 
        * **PR #16597** <a href="https://app.graphite.dev/github/pr/appium/appium/16597" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a> 

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

jlipps commented on packages/types/package.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "@appium/types",
+  "version": "0.0.0",
```
this will actually be published to NPM, correct? should we start at 0.0.1?

boneskull replied:
I figured we could just update that when publishing.

boneskull commented:
Build failure here is expected.

jlipps commented on packages/types/index.d.ts:
```diff
@@ -0,0 +1,610 @@
+import { Capabilities } from '@wdio/types';
```
is this file hand-rolled or auto-generated?

jlipps replied:
it's nice that we can reuse some of the work wdio has done, but internally we really don't care. internally it should just be:

```ts
{[key: string]: any}
```

maybe i've missed it and it's there somewhere already.

what if we add a new capability that wdio doesn't know about yet? what if a third party driver author adds a capability that wdio will never add to their list?

boneskull replied:
Then I'd say we should provide somewhere to define those things that do not occur in wdio's capabilities.

But in general, more specificity is better.  Any _optional_ properties should be typed, because they won't prevent someone from providing the wrong type, otherwise.  Further, you get inline documentation on the properties, where provided.

So I guess I'm in disagreement about using `any` here.  Even if we did avoid reusing wdio's types, there's still stuff that is absolutely required that both consumers and contributors should be aware of (e.g. W3C caps have a certain shape, `platformName` is required, etc.).  It may be that at some point in the control flow we don't really care any more, but it matters for public APIs.

I think we can reuse the W3C caps in a way which gives us more flexibility.  First, we can redefine what, e.g., `AppiumCapabiltiies` are by doing something like:

```ts
// untested
import type {AppiumCapabilities as BaseAppiumCapabilities} from '@wdio/types/capabilities';

interface OtherCapabilities {
 // whatever we need to add
 'appium:foo': string
}

export type AppiumCapabilities = BaseAppiumCapabilities & OtherCapabilities;
```

We can also use a generic to allow drivers to optionally specify their own capabilities:

```ts
export interface Driver<C extends AppiumCapabilities = AppiumCapabilities> {
  validateDesiredCaps(caps: C): boolean;
  // etc
}
```

An implementing class could then do something like:

```ts
interface MyCaps extends AppiumCapabilities {
  'stuff': 'things'
}

class MyMainClassDriver extends BaseDriver implements Driver<MyCaps> {

}
```

Of course, minding the types is completely optional.

What do you think?

jlipps replied:
Maybe we should put some effort into defining our own capability types then. I'm just not confident that WDIO's current ones are correct or will remain so even if they are.

Another issue to consider is that when we deal with capabilities internally, they don't have `appium:` in front, because this is stripped out in capabilities processing, but when users send capabilities to an Appium server (in which case they probably shouldn't be interfacing with these type definitions because they'll be using a client anyway, like WDIO, and consuming that client's type interface instead of ours), they do.

If we want to take this on, we'll need to hunt through the code to find the following types of capabilities mentioned so we can do the right thing with them:

- capabilities used by BaseDriver
- capabilities used by AppiumDriver
- capabilities used by a platform driver (which should be moved to a type definition in the platform driver instead)

And yes, in our internal code we'll always need to assume that the capabilities object could have other random things that the user has sent in. There are probably very few capabilities that we can definitely assume will always be present (I think platformName, automationName might be the only ones we can universally assume).

boneskull replied:
I've come up with something I think will be workable here.  `Capabilities` for internal use, and `W3CCapabilities` for external use.  We build these types off of wdio's types, but they are very loose, and we are free to extend them as we wish.  This provides autocompletion and _only_ errors if the type is actually _wrong_ (e.g., a `number` when a `string` was expected).  `W3CCapabilities` must correspond (roughly) to the standard.

boneskull commented:
(@jlipps 's review still says "changes requested" but this was approved otherwise via slack msg)

boneskull committed:
feat(appium): generate declaration files (#16597)

* feat(appium): generate declaration files

Configure TS to generate declaration files for `appium` package.

Most of these changes are just types getting added or changed via docstrings; in other places, typechecking has caught problems and I had to refactor code to address them.

- Removed `// @ts-check` pragma from all files in `lib/`, since now `checkJs: true` is in `tsconfig.json`.  It needs to stay in place elsewhere (I think) because only sources in `lib` have declarations generated.
- The `install`, `uninstall`, `update` and `run` functions in the base `ExtensionCommand` class were renamed and prefixed with a `_`; TS does not allow overridden functions with differing signatures.   It no longer exports a default.
- Removed `type` prop of options for `DriverCommand` and `PluginCommand` constructors, respectively (this value is derived from the `ExtensionConfig` `config` parameter instead)
- `AppiumDriver` now extends `DriverCore` instead of `BaseDriver`, since some of its function signatures differ from `BaseDriver`'s.
- Small refactor of `deleteSession` to make typechecker happy.
- `executeCommand` now calls to `BaseDriver` directly if `isUmbrellaCmd` is truthy.  This is one place where inheritance just kind of breaks, since `BaseDriver#executeCommand` returns something entirely different than `AppiumDriver#executeCommand` would otherwise.
- Removed `--delete-dir-on-start` from build command, since it may blast generated declarations.
- Reorganize declarations in `types/`; moved some of them out of `@typedef`s. `types.d.ts` becomes `index.d.ts`
- Note that this contains the declarations for `PluginClass` and `DriverClass`, which correspond to the types of extension "main" classes. This differs from the interfaces provided in `@appium/types` (e.g., `Driver`, `ExternalDriver`) in that these aren't just objects; they are classes with constructors and static members. I may end up moving some of this stuff into `@appium/types`, but I'm not sure how useful it will be.  Will see when I add types to `@appium/fake-driver`.

* chore(appium): remove transpile directives
this was ostensibly for gulp-babel

* chore(appium): remove redundant type declaration

* fix(appium): avoid potential unhandled rejections

- make types more readable

* chore(appium): fix capabilities types

- move `scripts/generate*` elsewhere
- add reference to `@appium/types` in `tsconfig.json`

* chore(appium): review updates for legibility

boneskull commented:
will merge these four PRs into 2.0 imminently (once build passes)

jlipps commented:
\o/ woohoo!
